File: AlgorithmDataStructure/LeetCode/36-Sudoku/sudoku.swift

class Solution {
    
    func isValidSudoku(_ board: [[String]]) -> Bool {
        var row = Array(repeating: Array(repeating: false, count: 9), count: 9)
        var col = Array(repeating: Array(repeating: false, count: 9), count: 9)
        var block = Array(repeating: Array(repeating: false, count: 9), count: 9)
        
        for i in 0 ..< 9 {
            for j in 0 ..< 9 {
                if board[i][j] != "." {
                    let num = Int("\(board[i][j])")! - 1
                    let blockIndex = i / 3 * 3 + j / 3
                    
                    if row[i][num] || col[j][num] || block[blockIndex][num] {
                        return false
                    } else {
                        row[i][num] = true
                        col[j][num] = true
                        block[blockIndex][num]  = true
                    }
                }
            }
        }
        return true
    }
    
}

let s = Solution()
var a = [["5","3",".",".","7",".",".",".","."],["6",".",".","1","9","5",".",".","."],[".","9","8",".",".",".",".","6","."],["8",".",".",".","6",".",".",".","3"],["4",".",".","8",".","3",".",".","1"],["7",".",".",".","2",".",".",".","6"],[".","6",".",".",".",".","2","8","."],[".",".",".","4","1","9",".",".","5"],[".",".",".",".","8",".",".","7","9"]]
print(s.isValidSudoku(a))

File: AlgorithmDataStructure/LeetCode/36-Sudoku/sudoku.c

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>


bool isRepeat(char *chars) {
    int marks[9] = {0};
    for (int i = 0; i < 9; i++) {
        char c = chars[i];
        if (c >= '1' && c <= '9') {
            int cInt = c - '1';
            if (marks[cInt] == 1) {
                return true;
            } else {
                marks[cInt] = 1;
            }
        }
    }
    return false;
}

bool isValidSudoku(char** board, int boardSize, int* boardColSize){
    if (boardSize <= 0) { return false; }
    
    //行
    for (int i = 0; i < 9; i++) {
        bool repeated = isRepeat(board[i]);
        if (repeated) {
            return false;
        }
    }
    
    //列
    for (int i = 0; i < 9; i++) {
        char *col = (char *)malloc(sizeof(char) * 9);
        for (int j = 0; j < 9; j++) {
            col[j] = board[j][i];
        }
        bool repeated = isRepeat(col);
        if (repeated) {
            return false;
        }
    }
    
    //方
    for (int n = 0; n < 3; n++) {
        for (int i = 0; i < 3; i++) {
            char *col = (char *)malloc(sizeof(char) * 9);
            for (int j = 0; j < 9; j++) {
                col[j] = board[i*3+j/3][n*3+j%3];
            }
            bool repeated = isRepeat(col);
            if (repeated) {
                return false;
            }
        }
    }
    
    return true;
}

int main(int argc, const char * argv[]) {
    
    for (int n = 0; n < 3; n++) {
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 9; j++) {
                printf("%d = %d %d \n", j, i*3+j/3, n*3+j%3);
            }
            printf("\n");
        }
        printf("------\n");
    }
    
    
    return 0;
}



